File: erlang/seminars/avl/src/bst.erl
-module(bst).

-export([tree/0, insert/3, lookup/2, depth/2, max_depth/1]).

tree() ->
    nil.

insert(nil, Key, Value) ->
    %% empty tree
    {node, Key, Value, nil, nil};

insert({node, Key, _, A, B}, Key, Value) ->
    {node, Key, Value, A, B};

insert({node, Rk, Rv, A, B}, Kk, Kv)  when Kk < Rk ->
    {node, Rk, Rv, insert(A, Kk, Kv), B};
insert({node, Rk, Rv, A, B}, Kk, Kv)  when Kk > Rk ->
    {node, Rk, Rv, A, insert(B, Kk, Kv)}.



lookup(nil, _) ->
    fail;
lookup({node, K, V,  _, _}, K) ->
    {ok, V};
lookup({node, K, _,  L, R}, Key) ->
    if
	Key < K -> lookup(L, Key);
	true -> lookup(R, Key)
    end.


traverse(nil) ->
    {0,0};
traverse({node, _, _, L, R}) ->
    {Lt, Lm} = traverse(L),
    {Rt, Rm} = traverse(R),    
    {Lt+Rt+1, max(Lm, Rm)+1}.


depth(nil, _) ->
    fail;
depth({node, K, _, _, _}, K) ->
    {ok, 1};
depth({node, K, _, L, R}, Key)  ->
    if
	Key < K -> 
	    case depth(L, Key) of
		{ok, D} -> {ok, D + 1};
		fail -> fail
	    end;
	true -> 
	    case depth(R, Key) of
		{ok, D} -> {ok, D + 1};
		fail -> fail
	    end
    end.



max_depth(nil) ->
    0;
max_depth({node, _, _, L, R})  ->
    max(max_depth(L), max_depth(R)) + 1.

File: exercises/avl/src/bst.ex
defmodule BST do

  # ================================================== #
  # == Binary Search Tree - Public API == #
  # ================================================== #

  # Create an empty tree.
  def tree(), do: nil

  # Insert a key-value pair in a tree.
  # Usage: insert(tree, key, value)
  def insert(nil, key, value) do
    {:node, key, value, nil, nil}
  end
  def insert({:node, key, _, a, b}, key, value) do
    {:node, key, value, a, b}
  end
  def insert({:node, rk, rv, a, b}, kk, kv) when kk < rk do
    {:node, rk, rv, insert(a, kk, kv), b}
  end
  def insert({:node, rk, rv, a, b}, kk, kv) when kk > rk do
    {:node, rk, rv, a, insert(b, kk, kv)}
  end

  # Lookup for a given key in a tree.
  # Usage: lookup(tree, key)
  def lookup(nil, _), do: :fail
  def lookup({:node, k, v, _, _}, k), do: {:ok, v}
  def lookup({:node, k, _, l, r}, key) do
    cond do
      key < k -> lookup(l, key)
      true -> lookup(r, key)
    end
  end

  # Compute the depth of a the leaf containing a given key
  # in a tree.
  # Usage: depth(tree, key)
  def depth(nil, _), do: :fail
  def depth({:node, k, _, _, _}, k), do: {:ok, 1}

  def depth({:node, k, _, l, r}, key) do
    cond do
      key < k ->
        case depth(l, key) do
          {:ok, d} -> {:ok, d + 1}
          :fail -> :fail
        end

      true ->
        case depth(r, key) do
          {:ok, d} -> {:ok, d + 1}
          :fail -> :fail
        end
    end
  end

  # Compute the maximum depth of a tree.
  # Usage: max_depth(tree)
  def max_depth(nil), do: 0
  def max_depth({:node, _, _, l, r}) do
    max(max_depth(l), max_depth(r)) + 1
  end

  # Traverse the tree to compute a tuple containing the 
  # number of nodes and the maximum depth.
  # Usage: traverse(tree)
  def traverse(nil), do: {0, 0}
  def traverse({:node, _, _, l, r}) do
    {lt, lm} = traverse(l)
    {rt, rm} = traverse(r)
    {lt + rt + 1, max(lm, rm) + 1}
  end
end
